java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java

```diff
@@ -145,6 +145,10 @@ private JavaType substituteInWildCardType(JavaType.WildCardType wildcard, TypeSu
     if (substitutedType != null) {
       return parametrizedTypeCache.getWildcardType(substitutedType, wildcard.boundType);
     }
+    JavaType newBound = applySubstitution(wildcard.bound, substitution);
+    if (newBound != wildcard.bound) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)